### PR TITLE
Support cloning of codelists

### LINKED ISSUE #2656: Make it easier to create a new version or copy of someone else's codelist
# Why are we doing this?

Currently it is not possible to directly copy or create a new version of a codelist that you don't own. However users want to do this, and so work round it by downloading the codes from the existing list, and creating a new version by uploading the CSV. But doing this means you lose the searches from the original codelist, and the metadata.

Making it easier to copy codelists may not be desirable due to the proliferation this might cause. However if we do not provide a way to update someone else's codelist then people will duplicate them anyway (via the download/upload CSV), and at least by providing users with a "Create copy" button we can copy across the searches, the metadata, and potentially a link back to the original.

# How will we know when it's done?

We have a better way to copy an existing codelist.

# What are we doing?

- [ ] Add a button on a published codelist to "Copy this codelist"
- [ ] This creates a new codelist, with the same searches, and include/excludes, but probably against the latest version of the coding system
- [ ] The metadata should be updated to link back to the original. This could just be as part of the "Description" field. 

---

[Defining delivery tasks guidance](https://bennett.wiki/tech-group/delivery-process/task-definition/)

### PR DESCRIPTION
**Pull request by Jongmassey**
Fixes #2656

### CODE REVIEW

File: codelists/actions.py [hunk 1]
```diff
@@ -746,3 +749,63 @@ def convert_bnf_codelist_version_to_dmd(version, published=False):
...
+        related = deepcopy(related)
+        version_field = None
+        for related_model_field in related.__class__._meta.get_fields():
+            if (
+                hasattr(related_model_field, "related_model")
+                and related_model_field.related_model == CodelistVersion
+            ):
+                version_field = related_model_field
+                break
+        assert version_field
```
Comment (rw251) [inline codelists/actions.py:806]: What was it you were saying the other day about `assert`s in the code...?

**Comment by rw251** [comment]
This looks good. A few comments attached. The only other thing is I'm wondering if we definitely want this to appear on a user's own codelists. Having "create new version" and "create fork" might confuse people, and I wonder if there is a danger that people end up with forks when they should have had a new version. Also this feature was a request specifically for the case of updating a codelist that a user didn't own - so I'm not sure there's a need for it. If we don't allow it, then we could always enable it later if people asked for it.

File: codelists/actions.py [hunk 2]
```diff
@@ -746,3 +749,63 @@ def convert_bnf_codelist_version_to_dmd(version, published=False):
...
+        setattr(related, version_field.name, forked_version)
+        related.id = None
+        related.save()
+
+    for field in CodelistVersion._meta.get_fields():
+        match field.__class__:
+            case reverse_related.OneToOneRel:
+                if hasattr(latest_version, field.name):
+                    related = getattr(latest_version, field.name)
+                    deep_copy_related(related)
```
[the remark this answers is not in the document]
Comment (Jongmassey) [reply codelists/actions.py:816]: Good point, it's not robust to more than one relational hop. I'll see if I can rewrite this to recurse the whole relation tree to make it robust.

File: codelists/views/codelist_fork.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+
+@require_GET
+@load_codelist
+@login_required
+def codelist_fork(request, codelist):
+    try:
+        forked_codelist = actions.fork_codelist(codelist, request.user)
+    except ValueError as e:
+        assert "Only user-owned codelists can be forked" in str(e)
```
[the remark this answers is not in the document]
Comment (Jongmassey) [reply codelists/views/codelist_fork.py:17]: Same (rubbish) excuse as the other `assert`, now replaced with something better!

File: codelists/tests/test_actions.py [hunk 4]
```diff
@@ -612,3 +613,82 @@ def test_can_update_converted_dmd_codelist(bnf_version_asthma, dmd_bnf_mapping_d
...
+    )
+    assert set(r.url for r in new_style_codelist.references.all()) == set(
+        r.url for r in forked_codelist.references.all()
+    )
+
+    # check forked codelist version
+    assert forked_codelist.versions.count() == 1
+    forked_version = forked_codelist.versions.first()
+    assert forked_version.status == Status.UNDER_REVIEW
+
```
[the remark this answers is not in the document]
Comment (Jongmassey) [reply codelists/tests/test_actions.py:737]: :+1:

File: codelists/actions.py [hunk 5]
```diff
@@ -761,9 +761,21 @@ def fork_codelist(codelist, new_owner):
         f"Forked from codelist : [{codelist.name}]({latest_version.get_absolute_url()}) \n"
         + (codelist.methodology or "")
     )
+    name = codelist.name
+    if codelist.user == new_owner:
+        n = 0
+        while True:
+            try:
+                new_name = f"{codelist.name}_copy{str(n) if n > 0 else ''}"
```
[the remark this answers is not in the document]
Comment (Jongmassey) [reply codelists/actions.py:808]: consistent naming is good

File: codelists/actions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jongmassey) [reply codelists/actions.py]: I _think_ I've nobbled it.

**Comment by Jongmassey** [comment]
image
Example of correct searches/results following fork

File: codelists/actions.py [hunk 6]
```diff
@@ -832,34 +832,68 @@ def fork_codelist(codelist, new_owner):
...
-                    deep_copy_related(related)
-            case reverse_related.ManyToOneRel:
-                if hasattr(latest_version, field.name):
-                    for related in getattr(latest_version, field.name).all():
-                        deep_copy_related(related)
-            case reverse_related.ManyToManyRel:
-                raise NotImplementedError
+    forked_related_objects = {}
+
+    def deep_copy_related_fields(source, destination):
```
Comment (rw251) [inline codelists/actions.py:837]: This looks good - but it's pretty complicated and I'm wondering if there may be unconsidered edge cases. If you're ok to own it I'm happy to go with it, but an alternative might be to explicitly copy all the relations that currently exist, and then add a test that fails if new relations are added to the model.

Comment (Jongmassey) [reply codelists/actions.py:837]: I'm happy to own this, the recursion is a bit ungainly due to the need to save the parent with a new ID before iterating the children but otherwise I'm happy. The tests give me enough confidence to proceed

**Review by rw251 — APPROVED** [review]

File: codelists/scripts/gbd_icd10.json
[the remark this answers is not in the document]
Comment (Jongmassey) [reply codelists/scripts/gbd_icd10.json:1]: D'oh yes
